File: app/src/main/java/pt/simdea/guestlist/Item.java
package pt.simdea.guestlist;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.objectbox.annotation.Entity;
import io.objectbox.annotation.Id;
import pt.simdea.gmlrva.lib.IGenericRecyclerViewLayout;
import pt.simdea.gmlrva.lib.IViewHolder;

@Entity
public class Item implements IGenericRecyclerViewLayout<Item.RowViewHolder> {
    @Id
    private long id;
    private int icon;
    private String item;
    private int counter;
    private boolean buyed;

    public Item() {
        this(-1, "", 0);
    }

    public Item(String title, int counter) {
        this(-1,title,counter);
    }

    public Item(int icon, String title, int counter) {
        this.icon = icon;
        this.item = title;
        this.counter = counter;
        buyed = false;
    }

    public int getIcon() {
        return icon;
    }

    public long getCounter() {
        return counter;
    }

    public void addCount() {
        counter++;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getItem() {
        return item;
    }

    public void setItem(String item) {
        this.item = item;
    }

    public boolean getBuyed() {
        return buyed;
    }

    public void setBuyed(boolean buyed) {
        this.buyed = buyed;
    }

    // Will be used by the ArrayAdapter in the ListView
    @Override
    public String toString() {
        return item;
    }

    @NonNull
    @Override
    public RowViewHolder createViewHolder(@NonNull ViewGroup parent) {
        return new RowViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.row, parent, false));
    }

    @Override
    public void setElements(@NonNull RowViewHolder holder) {
        holder.titleView.setText(item);
        holder.counterView.setText(String.format("%d", counter));
    }

    @NonNull
    @Override
    public Object getTag() {
        return item;
    }

    @Override
    public int getViewType() {
        return 0;
    }

    public class RowViewHolder extends RecyclerView.ViewHolder implements IViewHolder {

        @BindView(R.id.textView1)
        protected TextView titleView;
        @BindView(R.id.item_counter)
        protected TextView counterView;

        public RowViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        @Override
        public void recycle() {

        }
    }
}

File: app/src/main/kotlin/pt/simdea/guestlist/ui/Item.kt
package pt.simdea.guestlist

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import io.objectbox.annotation.Entity
import io.objectbox.annotation.Id
import pt.simdea.gmlrva.lib.IGenericRecyclerViewLayout
import pt.simdea.gmlrva.lib.IViewHolder
import pt.simdea.guestlist.databinding.RowBinding

@Entity
class Item @JvmOverloads constructor(internal val icon: Int = -1, internal var item: String? = "", private var counter: Int = 0) : IGenericRecyclerViewLayout<Item.RowViewHolder> {
    @Id
    public var id: Long = 0
    internal var buyed: Boolean = false

    constructor(title: String, counter: Int) : this(-1, title, counter)

    fun getCounter(): Long = counter.toLong()

    fun addCount() {
        counter++
    }

    override fun toString(): String = item!!

    override fun createViewHolder(parent: ViewGroup): RowViewHolder = RowViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.row, parent, false))

    override fun setElements(holder: RowViewHolder) {
        val binding = holder.dataBinding ?: return

        binding.titleView.text = item
        binding.counterView.text = String.format("%d", counter)
    }

    override fun getTag(): Any = item!!

    override fun getViewType(): Int = 0

    inner class RowViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), IViewHolder {
        val dataBinding: RowBinding? = DataBindingUtil.bind(itemView)

        override fun recycle() { }
    }

    fun getItem() = item
    fun getBuyed() = buyed
    fun getIcon() = icon
}
